README.md
# InterFAX notification channel for Laravel 9.x, 10.x

[![Latest Version on Packagist](https://img.shields.io/packagist/v/laravel-notification-channels/interfax.svg?style=flat-square)](https://packagist.org/packages/laravel-notification-channels/interfax)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/laravel-notification-channels/interfax/main.svg?style=flat-square)](https://travis-ci.org/laravel-notification-channels/interfax)
[![Total Downloads](https://img.shields.io/packagist/dt/laravel-notification-channels/interfax.svg?style=flat-square)](https://packagist.org/packages/laravel-notification-channels/interfax)

This package makes it easy to send notifications using [InterFAX](https://interfax.net) with Laravel 9.x and 10.x.

## Contents

- [Installation](#installation)
  - [Setting up the InterFAX service](#setting-up-the-InterFAX-service)
- [Usage](#usage)
  - [Available Message methods](#available-message-methods)
- [Changelog](#changelog)
- [Testing](#testing)
- [Security](#security)
- [Contributing](#contributing)
- [Credits](#credits)
- [License](#license)


## Installation

You can install this package via composer:

```bash
composer require laravel-notification-channels/interfax
```

The service provider gets loaded automatically.

### Setting up the InterFAX service

This channel will use your InterFAX username and password. To use the channel, add this to your `config/services.php` file:

```php
...
'interfax' => [
    'username' => env('INTERFAX_USERNAME'),
    'password' => env('INTERFAX_PASSWORD'),
    'pci'      => env('INTERFAX_PCI', false),
    'interval' => 15,
    'chunk_size' => null,
],
...
```

This will load your InterFAX credentials from the `.env` file. If your requests must be PCI-DSS-compliant, set `INTERFAX_PCI=true` in your `.env` file.

The `services.interfax.interval` configuration setting is the polling interval, in seconds, for a fax if it is set to check the status until it is complete. This is optional and will default to 15 if left empty. The interval has a minimum of 10 seconds, as the outbound service in the API has a maximum freqncy of 6 requests per minute and can return errors if polled more frequently.

Faxes can sometimes take more than 10 minutes to send, so it is recommended to configure a long-running queue and to push your fax notifications to that queue. More information on configuring long-running queues can be found [here](https://medium.com/@williamvicary/long-running-jobs-with-laravel-horizon-7655e34752f7).

The `services.interfax.chunk_size` configuration setting is the maximum file size before the InterFAX core SDK starts chunking files. The default chunk size is 1048576. When chunking, an `\Interfax\Document` object is created, but the `/outbound/documents` endpoint does not exist for the PCI-DSS-compliant API. If `services.interfax.pci` is set to `true`, it is recommended to increase the chunk size to avoid 404 errors.

## Usage

To use this package, you can create a notification class, like `DocumentWasSent` from the example below, in your Laravel application. Make sure to check out [Laravel's documentation](https://laravel.com/docs/master/notifications) for this process.

### Send PDF via fax

```php
<?php
use NotificationChannels\Interfax\InterfaxChannel;
use NotificationChannels\Interfax\InterfaxMessage;
use NotificationChannels\Interfax\Contracts\InterfaxNotificationContract;

class DocumentWasSent extends Notification implements InterfaxNotificationContract
{

    protected $files;

    public function __construct(array $files)
    {
        $this->files = $files;
    }

    /**
     * Get the notification's delivery channels.
     *
     * @param  mixed  $notifiable
     * @return array
     */
    public function via($notifiable)
    {
        return [InterfaxChannel::class];
    }

    public function toInterfax($notifiable): InterfaxMessage
    {
        return (new InterfaxMessage)
              ->files($this->files);
    }
}
```

The Notifiable model will need to return a destination fax number.

```php
public function routeNotificationForInterfax($notification)
{
    if($this->fax)
        return preg_replace('/[^\d]/', '', $this->fax);

    return null;
}
```

### Available Message methods

`file(string $file)` : Accepts the full path to a single file (full list of supported file types [found here](https://www.interfax.net/en/help/supported_file_types)).  
`files(array $array)` : Accepts an array of file paths.  If overriding the default chunk_size in the config and using an `\Interfax\File` object in the array, use `\NotificationChannels\Interfax\InterfaxFile` instead to automatically set the file's chunk size on initialization.  
`stream(Filestream $stream, string $name)` : Accepts a file stream.  
`addMetadata(array $data)`: Add metadata for logging purposes in case of an error.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Security

If you discover any security related issues, please email craig.spivack@ivinteractive.com instead of using the issue tracker.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [Craig Spivack](https://github.com/iv-craig)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to the `InterFAX notification channel` will be documented in this file.

## 3.0.0 - 2023-02-24

- Add Laravel 10 support
- Drop support for EOL versions of PHP (7.2-7.4) & Laravel (6.x, 7.x, 8.x)
- Add stronger typing based on recommendations from PHPStan

## 2.4.0 - 2022-04-12

- Add Laravel 9 support

## 2.3.0 - 2022-01-18

- Add functionality for overriding default file size where fax documents start chunking

## 2.2.0 - 2022-01-04

- Add support for PHP 8.0 & 8.1
- Improve error handling
- Add `addMetadata` method to add array of data to be used for error logging purposes

## 2.1.1 - 2021-05-12

- Prevent API limit requests with updates to status polling
- Update exception handling for better error messages from InterFAX client

## 2.1.0 - 2020-11-10

- Add Laravel 8 support

## 2.0.0 - 2020-07-15

- Add support for Laravel 7.x
- Drop support for Laravel 5.8
- Add package autodiscovery

## 1.0.0 - 2020-01-08

- Initial release


tests/TestCase.php
<?php

namespace NotificationChannels\Interfax\Test;

use Mockery;

abstract class TestCase extends \Orchestra\Testbench\TestCase
{
    protected $testFiles = [
        'test-file.pdf',
        'test-file-1.pdf',
        'test-file-2.pdf',
    ];

    protected $chunkSize = 1048576;

    protected function getPackageProviders($app)
    {
        config([
            'services.interfax.username' => 'user',
            'services.interfax.password' => 'pass',
        ]);

        return [\NotificationChannels\Interfax\InterfaxServiceProvider::class];
    }

    public function setUp(): void
    {
        parent::setUp();

        $mpdf = new \Mpdf\Mpdf;
        $mpdf->WriteHTML('<h1>Test file contents</h1>');

        foreach ($this->testFiles as $file) {
            $mpdf->Output(app('filesystem')->path($file), \Mpdf\Output\Destination::FILE);
        }
    }

    public function tearDown(): void
    {
        foreach ($this->testFiles as $file) {
            app('filesystem')->delete($file);
        }

        parent::tearDown();

        if ($container = \Mockery::getContainer()) {
            $this->addToAssertionCount($container->mockery_getExpectationCount());
        }

        Mockery::close();
    }

    protected function addFile(string $filename): void
    {
        $this->testFiles[] = $filename;
    }

    protected function increaseChunkSize(int $size = 8000000)
    {
        $this->chunkSize = $size;
        config(['services.interfax.chunk_size' => $this->chunkSize]);
    }
}

class TestNotifiable
{
    use \Illuminate\Notifications\Notifiable;

    /**
     * @return string
     */
    public function routeNotificationForInterfax()
    {
        return '12345678901';
    }
}

class TestNotifiableNotSendable
{
    use \Illuminate\Notifications\Notifiable;

    /**
     * @return bool
     */
    public function routeNotificationForInterfax()
    {
        return false;
    }
}


tests/InterfaxChannelTest.php
<?php

namespace NotificationChannels\Interfax\Test;

use Illuminate\Notifications\Notification;
use Interfax\Client as InterfaxClient;
use Interfax\Resource as InterfaxResource;
use Mockery;
use NotificationChannels\Interfax\Exceptions\CouldNotSendNotification;
use NotificationChannels\Interfax\Contracts\InterfaxNotificationContract;
use NotificationChannels\Interfax\InterfaxChannel;
use NotificationChannels\Interfax\InterfaxMessage;

class InterfaxChannelTest extends TestCase
{
    /** @var Mockery\Mock */
    protected $interfax;

    /** @var Mockery\Mock */
    protected $resource;

    /** @var \NotificationChannels\Interfax\InterfaxChannel */
    protected $channel;

    public function setUp(): void
    {
        parent::setUp();

        config([
            'services.interfax.interval' => 1,
        ]);

        $this->interfax = Mockery::mock(InterfaxClient::class);
        $this->resource = Mockery::mock(InterfaxResource::class);
        $this->channel = new InterfaxChannel($this->interfax);
    }

    /** @test */
    public function it_can_send_notification_with_a_single_file()
    {
        $this->interfax->expects('deliver')
            ->once()
            ->with([
                'faxNumber' => '12345678901',
                'files' => [
                    [
                        'location' => 'test-file.pdf',
                        'params' => [
                            'chunk_size' => $this->chunkSize,
                        ],
                    ],
                ],
            ]);

        $this->channel->send(new TestNotifiable, new TestNotificationWithSingleFile);
    }

    /** @test */
    public function it_can_send_notification_with_files()
    {
        $this->interfax->expects('deliver')
            ->once()
            ->with([
                'faxNumber' => '12345678901',
                'files' => [
                    [
                        'location' => 'test-file-1.pdf',
                        'params' => [
                            'chunk_size' => $this->chunkSize,
                        ],
                    ],
                    [
                        'location' => 'test-file-2.pdf',
                        'params' => [
                            'chunk_size' => $this->chunkSize,
                        ],
                    ],
                ],
            ]);

        $this->channel->send(new TestNotifiable, new TestNotificationWithFiles);
    }

    /** @test */
    public function it_can_send_notification_pdf_as_stream()
    {
        $this->interfax->expects('deliver')
            ->with(Mockery::on(function ($output) {
                if ($output['faxNumber'] !== '12345678901') {
                    return false;
                }

                $file = $output['files'][0];

                if ($file[1]['name'] !== app('filesystem')->path('test-file.pdf')) {
                    return false;
                }

                if ($file[1]['mime_type'] !== 'application/pdf') {
                    return false;
                }

                if (! is_resource($file[0])) {
                    return false;
                }

                return true;
            }));

        $this->channel->send(new TestNotifiable, new TestNotificationAsStreamPdf);
    }

    /** @test */
    public function it_can_send_notification_html_as_stream()
    {
        $filename = 'test-file.html';
        $this->addFile($filename);

        app('filesystem')->put($filename, '<html><body><h1>Test file contents</h1></body></html>');

        $this->interfax->expects('deliver')
            ->with(Mockery::on(function ($output) {
                if ($output['faxNumber'] !== '12345678901') {
                    return false;
                }

                $file = $output['files'][0];

                if ($file[1]['name'] !== app('filesystem')->path('test-file.html')) {
                    return false;
                }

                if ($file[1]['mime_type'] !== 'text/html') {
                    return false;
                }

                if (! is_resource($file[0])) {
                    return false;
                }

                return true;
            }));

        $this->channel->send(new TestNotifiable, new TestNotificationAsStreamHtml);
    }

    /** @test */
    public function it_can_return_early_when_no_fax_number_provided()
    {
        $this->assertNull($this->channel->send(new TestNotifiableNotSendable, new TestNotificationWithFiles));
    }

    /** @test */
    public function it_can_refresh_the_file_response()
    {
        $this->resource
             ->expects('refresh')
             ->times(3)
             ->andReturn((object) [
                 'status' => -1,
             ], (object) [
                 'status' => 0,
             ]);

        $this->interfax
             ->expects('deliver')
             ->andReturn($this->resource);

        $this->channel->send(new TestNotifiable, new TestNotificationWithRefresh);
    }

    /** @test */
    public function it_can_throw_the_exception()
    {
        $this->expectException(CouldNotSendNotification::class);

        $testResource = new TestResource;

        $this->resource
             ->expects('refresh')
             ->times(2)
             ->andReturn($testResource);

        $this->resource
             ->expects('attributes')
             ->andReturn($testResource->attributes());

        $this->interfax
             ->expects('deliver')
             ->andReturn($this->resource);

        $this->channel->send(new TestNotifiable, new TestNotificationWithRefresh);
    }
}

class TestNotificationWithSingleFile extends Notification implements InterfaxNotificationContract
{
    /**
     * @param $notifiable
     * @return InterfaxMessage
     *
     * @throws CouldNotSendNotification
     */
    public function toInterfax($notifiable): InterfaxMessage
    {
        return (new InterfaxMessage)
                    ->user($notifiable)
                    ->file('test-file.pdf');
    }
}

class TestNotificationWithFiles extends Notification implements InterfaxNotificationContract
{
    /**
     * @param $notifiable
     * @return InterfaxMessage
     *
     * @throws CouldNotSendNotification
     */
    public function toInterfax($notifiable): InterfaxMessage
    {
        return (new InterfaxMessage)
                    ->user($notifiable)
                    ->files(['test-file-1.pdf', 'test-file-2.pdf']);
    }
}

class TestNotificationAsStreamPdf extends Notification implements InterfaxNotificationContract
{
    /**
     * @param $notifiable
     * @return InterfaxMessage
     *
     * @throws CouldNotSendNotification
     */
    public function toInterfax($notifiable): InterfaxMessage
    {
        $path = app('filesystem')->path('test-file.pdf');

        return (new InterfaxMessage)
                    ->user($notifiable)
                    ->stream(fopen($path, 'r'), $path);
    }
}

class TestNotificationAsStreamHtml extends Notification implements InterfaxNotificationContract
{
    /**
     * @param $notifiable
     * @return InterfaxMessage
     *
     * @throws CouldNotSendNotification
     */
    public function toInterfax($notifiable): InterfaxMessage
    {
        $path = app('filesystem')->path('test-file.html');

        return (new InterfaxMessage)
                    ->user($notifiable)
                    ->stream(fopen($path, 'r'), $path);
    }
}

class TestNotificationWithRefresh extends Notification implements InterfaxNotificationContract
{
    /**
     * @param $notifiable
     * @return InterfaxMessage
     *
     * @throws CouldNotSendNotification
     */
    public function toInterfax($notifiable): InterfaxMessage
    {
        return (new InterfaxMessage)
                    ->checkStatus()
                    ->user($notifiable)
                    ->files(['test-file-1.pdf', 'test-file-2.pdf']);
    }
}

class TestResource
{
    public $status = 500;

    public function attributes()
    {
        return [
            'status' => $this->status,
            'message' => 'Failed to send.',
        ];
    }
}


tests/InterfaxMessageTest.php
<?php

namespace NotificationChannels\Interfax\Test;

use NotificationChannels\Interfax\InterfaxMessage;

class InterfaxMessageTest extends TestCase
{
    /** @test */
    public function it_should_check_the_status_via_refresh()
    {
        $message = (new InterfaxMessage)
                        ->checkStatus()
                        ->user(new TestNotifiable)
                        ->file('test-file.pdf');

        $this->assertTrue($message->shouldCheckStatus());
    }

    /** @test */
    public function it_should_not_check_the_status_via_refresh_manual()
    {
        $message = (new InterfaxMessage)
                        ->checkStatus(false)
                        ->user(new TestNotifiable)
                        ->file('test-file.pdf');

        $this->assertFalse($message->shouldCheckStatus());
    }

    /** @test */
    public function it_should_not_check_the_status_via_refresh_default()
    {
        $message = (new InterfaxMessage)
                        ->user(new TestNotifiable)
                        ->file('test-file.pdf');

        $this->assertFalse($message->shouldCheckStatus());
    }

    /** @test */
    public function it_should_set_the_file_chunk_size_filename()
    {
        $this->increaseChunkSize();

        $message = (new InterfaxMessage)
                        ->user(new TestNotifiable)
                        ->file(__DIR__.'/resources/test.pdf');

        $files = $message->makeFiles();
        $delivery = new \Interfax\Outbound\Delivery(new \Interfax\Client, ['faxNumber'=>'0000000000', 'files'=>$files]);

        $this->assertSame($this->chunkSize, $this->getChunkSize($delivery));
    }

    /** @test */
    public function it_should_set_the_file_chunk_size_file_array()
    {
        $this->increaseChunkSize();

        $message = (new InterfaxMessage)
                        ->user(new TestNotifiable)
                        ->files([['location' => __DIR__.'/resources/test.pdf']]);

        $files = $message->makeFiles();
        $delivery = new \Interfax\Outbound\Delivery(new \Interfax\Client, ['faxNumber'=>'0000000000', 'files'=>$files]);

        $this->assertSame($this->chunkSize, $this->getChunkSize($delivery));
    }

    /** @test */
    public function it_should_set_the_file_chunk_size_file_object()
    {
        $this->increaseChunkSize();
        $client = new \Interfax\Client;

        $file = new \NotificationChannels\Interfax\InterfaxFile($client, __DIR__.'/resources/test.pdf');

        $message = (new InterfaxMessage)
                        ->user(new TestNotifiable)
                        ->files([$file]);

        $files = $message->makeFiles();
        $delivery = new \Interfax\Outbound\Delivery($client, ['faxNumber'=>'0000000000', 'files'=>$files]);

        $this->assertSame($this->chunkSize, $this->getChunkSize($delivery));
    }

    protected function getChunkSize($delivery)
    {
        $deliveryReflection = new \ReflectionClass($delivery);
        $filesProperty = $deliveryReflection->getProperty('files');
        $filesProperty->setAccessible(true);

        $files = $filesProperty->getValue($delivery);

        $fileReflection = new \ReflectionClass($files[0]);

        if ($files[0] instanceof \NotificationChannels\Interfax\InterfaxFile) {
            $fileReflection = $fileReflection->getParentClass();
        }

        $chunkProperty = $fileReflection->getProperty('chunk_size');
        $chunkProperty->setAccessible(true);

        return $chunkProperty->getValue($files[0]);
    }
}


src/InterfaxChannel.php
<?php

namespace NotificationChannels\Interfax;

use Interfax\Client;
use NotificationChannels\Interfax\Contracts\InterfaxNotificationContract;
use NotificationChannels\Interfax\Exceptions\CouldNotSendNotification;

class InterfaxChannel
{
    protected $client;
    protected $fax;

    public function __construct(Client $client)
    {
        $this->client = $client;
    }

    /**
     * Send the given notification.
     *
     * @param  mixed  $notifiable
     * @param  \NotificationChannels\Interfax\Contracts\InterfaxNotificationContract  $notification
     *
     * @throws \NotificationChannels\Interfax\Exceptions\CouldNotSendNotification
     */
    public function send($notifiable, InterfaxNotificationContract $notification)
    {
        if (! $faxNumber = $notifiable->routeNotificationFor('interfax')) {
            return;
        }

        $message = $notification->toInterfax($notifiable);

        try {
            $this->fax = $this->client->deliver([
                'faxNumber' => $faxNumber,
                'files' => $message->makeFiles(),
            ]);

            if ($message->shouldCheckStatus()) {
                $message->sleep();

                while ($this->fax->refresh()->status < 0) {
                    $message->sleep();
                }

                if ($this->fax->refresh()->status > 0) {
                    throw CouldNotSendNotification::serviceRespondedWithAnError($message, $this->fax->attributes());
                }
            }
        } catch (\Interfax\Exception\RequestException $e) {
            $exceptionMessage = $e->getMessage().': '.($e->getWrappedException())->getMessage();
            $attributes = $this->fax ? $this->fax->attributes() : ['status' => $e->getStatusCode()];

            throw CouldNotSendNotification::serviceRespondedWithAnError($message, $attributes, $exceptionMessage);
        }
    }
}


src/Exceptions/CouldNotSendNotification.php
<?php

namespace NotificationChannels\Interfax\Exceptions;

use Exception;
use NotificationChannels\Interfax\InterfaxMessage;

class CouldNotSendNotification extends Exception
{
    protected $interfaxMessage;
    protected $responseAttributes;

    final public function __construct(string $message, int $code, Exception $previous = null, InterfaxMessage $interfaxMessage, array $responseAttributes)
    {
        parent::__construct($message, $code, $previous);

        $this->interfaxMessage = $interfaxMessage;
        $this->responseAttributes = $responseAttributes;
    }

    public static function serviceRespondedWithAnError($message, $responseAttributes, string $exceptionMessage = 'The fax failed to send via InterFAX.')
    {
        return new static($exceptionMessage, $responseAttributes['status'], null, $message, $responseAttributes);
    }

    public function getUser()
    {
        return $this->interfaxMessage->user;
    }

    public function getMetadata()
    {
        return $this->interfaxMessage->metadata;
    }

    public function getAttributes()
    {
        return $this->responseAttributes;
    }
}


src/InterfaxMessage.php
<?php

namespace NotificationChannels\Interfax;

use Illuminate\Support\Arr;

class InterfaxMessage
{
    protected array $files;
    protected $stream;
    protected string $filename;
    protected string $method;
    protected $statusCheck = false;
    public $user;
    public array $metadata = [];

    const FILES = 'files';
    const STREAM = 'stream';

    const POLLING_INTERVAL_DEFAULT = 15;
    const POLLING_INTERVAL_MINIMUM = 10;

    protected static $DEFAULT_CHUNK_SIZE = 1048576;

    public function file(string $file)
    {
        $this->files = Arr::wrap($file);
        $this->method = static::FILES;

        return $this;
    }

    public function files(array $files): InterfaxMessage
    {
        $this->files = $files;
        $this->method = static::FILES;

        return $this;
    }

    public function stream($stream, string $filename): InterfaxMessage
    {
        $this->stream = $stream;
        $this->filename = $filename;
        $this->method = static::STREAM;

        return $this;
    }

    public function checkStatus(bool $shouldCheck = true)
    {
        $this->statusCheck = $shouldCheck;

        return $this;
    }

    public function shouldCheckStatus(): bool
    {
        return $this->statusCheck;
    }

    /**
     * Set a user who can be notified in case the fax fails to send.
     *
     * @param  mixed  $notifiable  The user to notify
     * @return InterfaxMessage
     */
    public function user($notifiable): InterfaxMessage
    {
        $this->user = $notifiable;

        return $this;
    }

    /**
     * Add metadata to the message for logging purposes.
     *
     * @param  array  $data  The data to add to the metadata array
     * @return InterfaxMessage
     */
    public function addMetadata(array $data): InterfaxMessage
    {
        foreach ($data as $key => $value) {
            $this->metadata[$key] = $value;
        }

        return $this;
    }

    public function makeFiles(): array
    {
        if ($this->method === static::STREAM) {
            return [
                [
                    $this->stream,
                    [
                        'name' => $this->filename,
                        'mime_type' => app('files')->mimeType($this->filename),
                        'chunk_size' => config('services.interfax.chunk_size', static::$DEFAULT_CHUNK_SIZE),
                    ],
                ],
            ];
        }

        return array_map('static::setChunkSize', $this->files);
    }

    public function sleep(): void
    {
        $interval = config('services.interfax.interval', static::POLLING_INTERVAL_DEFAULT);
        sleep(max($interval, static::POLLING_INTERVAL_MINIMUM));
    }

    protected static function setChunkSize($file)
    {
        $chunk_size = config('services.interfax.chunk_size', static::$DEFAULT_CHUNK_SIZE);

        if (is_string($file)) {
            return [
                'location' => $file,
                'params' => [
                    'chunk_size' => $chunk_size,
                ],
            ];
        } elseif (is_array($file)) {
            $file['params']['chunk_size'] = $chunk_size;

            return $file;
        } else {
            return $file;
        }
    }
}
